SAML AdvancedAttributeToRoleMapper does not allow predicate evaluation on same Array Attribute

### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

saml

### Describe the bug

![d3850847-10cc-464b-81a5-688daa2a5805](https://github.com/keycloak/keycloak/assets/15670023/2f576c5c-f01e-4db4-8d1e-6a2afbaf7d91)
```
<saml:Attribute xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Name="groupMembership" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
	<saml:AttributeValue xsi:type="xs:string">cn=captain,ou=...,ou=...,o=...</saml:AttributeValue>
	<saml:AttributeValue xsi:type="xs:string">cn=america,ou=...,ou=...,o=...</saml:AttributeValue>
```

Configuration above leads to an "duplicate key groupMembership..." error in Keycloak's log and no mapping at all.

### Version

22.0.1

### Expected behavior

I expect following behaviour (for the example above):

_if I am a group member of "america" and a group member of "captain" give  me the role "manager_us"_

### Actual behavior

Predicates (equals or regex) on different keys evaluate with AND.
Predicates on same key evaluate not at all.

### How to Reproduce?

Define predicates on same key.

### Anything else?

I consider this a bug because of the inconsistent behavior and the inability to express a common use-case.

Fix should be easy when I look into the source code.

```
2023-08-24 13:24:23,324 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-12) Uncaught server error: org.keycloak.broker.provider.IdentityBrokerException: Could not process response from SAML identity provider.
	at org.keycloak.broker.saml.SAMLEndpoint$Binding.handleLoginResponse(SAMLEndpoint.java:573)
	at org.keycloak.broker.saml.SAMLEndpoint$Binding.handleSamlResponse(SAMLEndpoint.java:657)
	at org.keycloak.broker.saml.SAMLEndpoint$Binding.execute(SAMLEndpoint.java:276)
	at org.keycloak.broker.saml.SAMLEndpoint.postBinding(SAMLEndpoint.java:187)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:430)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:408)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$0(QuarkusRequestFilter.java:82)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Duplicate key groupMembership (attempted merging values captain.* and america.*)
	at java.base/java.util.stream.Collectors.duplicateKeyException(Collectors.java:133)
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:180)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.keycloak.models.IdentityProviderMapperModel.getConfigMap(IdentityProviderMapperModel.java:101)
	at org.keycloak.broker.saml.mappers.AdvancedAttributeToRoleMapper.applies(AdvancedAttributeToRoleMapper.java:126)
	at org.keycloak.broker.saml.mappers.AbstractAttributeToRoleMapper.updateBrokeredUser(AbstractAttributeToRoleMapper.java:63)
	at org.keycloak.broker.provider.IdentityProviderMapperSyncModeDelegate.delegateUpdateBrokeredUser(IdentityProviderMapperSyncModeDelegate.java:25)
	at org.keycloak.services.resources.IdentityBrokerService.lambda$updateFederatedIdentity$3(IdentityBrokerService.java:1010)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.keycloak.services.resources.IdentityBrokerService.updateFederatedIdentity(IdentityBrokerService.java:1007)
	at org.keycloak.services.resources.IdentityBrokerService.authenticated(IdentityBrokerService.java:608)
	at org.keycloak.broker.saml.SAMLEndpoint$Binding.handleLoginResponse(SAMLEndpoint.java:569)
```

--- a/server-spi/src/main/java/org/keycloak/models/IdentityProviderMapperModel.java
+++ b/server-spi/src/main/java/org/keycloak/models/IdentityProviderMapperModel.java
@@ -22,6 +22,7 @@ import org.keycloak.util.JsonSerialization;
 
 import java.io.IOException;
 import java.io.Serializable;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
@@ -93,12 +94,16 @@ public class IdentityProviderMapperModel implements Serializable {
         this.config = config;
     }
 
-    public Map<String, String> getConfigMap(String configKey) {
+    public Map<String, List<String>> getConfigMap(String configKey) {
         String configMap = config.get(configKey);
 
         try {
             List<StringPair> map = JsonSerialization.readValue(configMap, MAP_TYPE_REPRESENTATION);
-            return map.stream().collect(Collectors.toMap(StringPair::getKey, StringPair::getValue));
+            return map.stream().collect(
+                    Collectors.collectingAndThen(
+                            Collectors.groupingBy(StringPair::getKey,
+                                    Collectors.mapping(StringPair::getValue, Collectors.toUnmodifiableList())),
+                            Collections::unmodifiableMap));
         } catch (IOException e) {
             throw new RuntimeException("Could not deserialize json: " + configMap, e);
         }

--- a/services/src/main/java/org/keycloak/broker/oidc/mappers/AdvancedClaimToGroupMapper.java
+++ b/services/src/main/java/org/keycloak/broker/oidc/mappers/AdvancedClaimToGroupMapper.java
@@ -100,15 +100,16 @@ public class AdvancedClaimToGroupMapper extends AbstractClaimToGroupMapper {
 
     @Override
     protected boolean applies(IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
-        Map<String, String> claims = mapperModel.getConfigMap(CLAIM_PROPERTY_NAME);
+        Map<String, List<String>> claims = mapperModel.getConfigMap(CLAIM_PROPERTY_NAME);
         boolean areClaimValuesRegex = Boolean.parseBoolean(mapperModel.getConfig().get(ARE_CLAIM_VALUES_REGEX_PROPERTY_NAME));
 
-        for (Map.Entry<String, String> claim : claims.entrySet()) {
-            Object value = getClaimValue(context, claim.getKey());
-
-            boolean claimValuesMismatch = !(areClaimValuesRegex ? valueMatchesRegex(claim.getValue(), value) : valueEquals(claim.getValue(), value));
-            if (claimValuesMismatch) {
-                return false;
+        for (Map.Entry<String, List<String>> claim : claims.entrySet()) {
+            Object claimValue = getClaimValue(context, claim.getKey());
+            for (String value : claim.getValue()) {
+                boolean claimValuesMismatch = !(areClaimValuesRegex ? valueMatchesRegex(value, claimValue) : valueEquals(value, claimValue));
+                if (claimValuesMismatch) {
+                    return false;
+                }
             }
         }
 

--- a/services/src/main/java/org/keycloak/broker/oidc/mappers/AdvancedClaimToRoleMapper.java
+++ b/services/src/main/java/org/keycloak/broker/oidc/mappers/AdvancedClaimToRoleMapper.java
@@ -46,7 +46,7 @@ public class AdvancedClaimToRoleMapper extends AbstractClaimToRoleMapper {
     public static final String[] COMPATIBLE_PROVIDERS = {KeycloakOIDCIdentityProviderFactory.PROVIDER_ID, OIDCIdentityProviderFactory.PROVIDER_ID};
     private static final Set<IdentityProviderSyncMode> IDENTITY_PROVIDER_SYNC_MODES = new HashSet<>(Arrays.asList(IdentityProviderSyncMode.values()));
 
-    private static final List<ProviderConfigProperty> configProperties = new ArrayList<ProviderConfigProperty>();
+    private static final List<ProviderConfigProperty> configProperties = new ArrayList<>();
 
     static {
         ProviderConfigProperty claimsProperty = new ProviderConfigProperty();
@@ -108,15 +108,16 @@ public class AdvancedClaimToRoleMapper extends AbstractClaimToRoleMapper {
 
     @Override
     protected boolean applies(IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
-        Map<String, String> claims = mapperModel.getConfigMap(CLAIM_PROPERTY_NAME);
+        Map<String, List<String>> claims = mapperModel.getConfigMap(CLAIM_PROPERTY_NAME);
         boolean areClaimValuesRegex = Boolean.parseBoolean(mapperModel.getConfig().get(ARE_CLAIM_VALUES_REGEX_PROPERTY_NAME));
 
-        for (Map.Entry<String, String> claim : claims.entrySet()) {
-            Object value = getClaimValue(context, claim.getKey());
-
-            boolean claimValuesMismatch = !(areClaimValuesRegex ? valueMatchesRegex(claim.getValue(), value) : valueEquals(claim.getValue(), value));
-            if (claimValuesMismatch) {
-                return false;
+        for (Map.Entry<String, List<String>> claim : claims.entrySet()) {
+            Object claimValue = getClaimValue(context, claim.getKey());
+            for (String value : claim.getValue()) {
+                boolean claimValuesMismatch = !(areClaimValuesRegex ? valueMatchesRegex(value, claimValue) : valueEquals(value, claimValue));
+                if (claimValuesMismatch) {
+                    return false;
+                }
             }
         }
 

--- a/services/src/main/java/org/keycloak/broker/oidc/mappers/ClaimToUserSessionNoteMapper.java
+++ b/services/src/main/java/org/keycloak/broker/oidc/mappers/ClaimToUserSessionNoteMapper.java
@@ -104,30 +104,31 @@ public class ClaimToUserSessionNoteMapper extends AbstractClaimMapper {
     }
 
     private void addClaimsToSessionNote(IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
-        Map<String, String> claims = mapperModel.getConfigMap(CLAIMS_PROPERTY_NAME);
+        Map<String, List<String>> claims = mapperModel.getConfigMap(CLAIMS_PROPERTY_NAME);
         boolean areClaimValuesRegex =
                 Boolean.parseBoolean(mapperModel.getConfig().get(ARE_CLAIM_VALUES_REGEX_PROPERTY_NAME));
 
-        for (Map.Entry<String, String> claim : claims.entrySet()) {
-            Object valueObj = getClaimValue(context, claim.getKey());
+        for (Map.Entry<String, List<String>> claim : claims.entrySet()) {
+            Object claimValueObj = getClaimValue(context, claim.getKey());
+            for (String value : claim.getValue()) {
+                if (claimValueObj != null) {
+                    if (!(claimValueObj instanceof String)) {
+                        LOG.warnf(
+                                "Claim '%s' does not contain a string value for user with brokerUserId '%s'. "
+                                + "Actual value is of type '%s': %s",
+                                claim.getKey(),
+                                context.getBrokerUserId(), claimValueObj.getClass(), claimValueObj);
+                        continue;
+                    }
 
-            if (valueObj != null) {
-                if (!(valueObj instanceof String)) {
-                    LOG.warnf(
-                            "Claim '%s' does not contain a string value for user with brokerUserId '%s'. " +
-                                    "Actual value is of type '%s': %s",
-                            claim.getKey(),
-                            context.getBrokerUserId(), valueObj.getClass(), valueObj);
-                    continue;
-                }
+                    String claimValue = (String) claimValueObj;
 
-                String value = (String) valueObj;
+                    boolean claimValuesMatch = areClaimValuesRegex ? valueMatchesRegex(value, claimValue)
+                            : valueEquals(value, claimValue);
 
-                boolean claimValuesMatch = areClaimValuesRegex ? valueMatchesRegex(claim.getValue(), value)
-                        : valueEquals(claim.getValue(), value);
-
-                if (claimValuesMatch) {
-                    context.getAuthenticationSession().setUserSessionNote(claim.getKey(), value);
+                    if (claimValuesMatch) {
+                        context.getAuthenticationSession().setUserSessionNote(claim.getKey(), claimValue);
+                    }
                 }
             }
         }

--- a/services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToGroupMapper.java
+++ b/services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToGroupMapper.java
@@ -17,7 +17,6 @@
 
 package org.keycloak.broker.saml.mappers;
 
-
 import org.keycloak.broker.provider.BrokeredIdentityContext;
 import org.keycloak.broker.provider.ConfigConstants;
 import org.keycloak.broker.saml.SAMLEndpoint;
@@ -114,8 +113,9 @@ public class AdvancedAttributeToGroupMapper extends AbstractAttributeToGroupMapp
         return "If all attributes exists, assign the user to the specified group.";
     }
 
+    @Override
     protected boolean applies(final IdentityProviderMapperModel mapperModel, final BrokeredIdentityContext context) {
-        Map<String, String> attributes = mapperModel.getConfigMap(ATTRIBUTE_PROPERTY_NAME);
+        Map<String, List<String>> attributes = mapperModel.getConfigMap(ATTRIBUTE_PROPERTY_NAME);
         boolean areAttributeValuesRegexes = Boolean.parseBoolean(mapperModel.getConfig().get(ARE_ATTRIBUTE_VALUES_REGEX_PROPERTY_NAME));
 
         AssertionType assertion = (AssertionType) context.getContextData().get(SAMLEndpoint.SAML_ASSERTION);
@@ -124,19 +124,21 @@ public class AdvancedAttributeToGroupMapper extends AbstractAttributeToGroupMapp
             return false;
         }
 
-        for (Map.Entry<String, String> attribute : attributes.entrySet()) {
-            String attributeKey = attribute.getKey();
-            List<Object> attributeValues = attributeAssertions.stream()
-                    .flatMap(statements -> statements.getAttributes().stream())
-                    .filter(choiceType -> attributeKey.equals(choiceType.getAttribute().getName())
-                            || attributeKey.equals(choiceType.getAttribute().getFriendlyName()))
-                    // Several statements with same name are treated like one with several values
-                    .flatMap(choiceType -> choiceType.getAttribute().getAttributeValue().stream())
-                    .collect(Collectors.toList());
+        for (Map.Entry<String, List<String>> entry : attributes.entrySet()) {
+            String attributeKey = entry.getKey();
+            for (String value : entry.getValue()) {
+                List<Object> attributeValues = attributeAssertions.stream()
+                        .flatMap(statements -> statements.getAttributes().stream())
+                        .filter(choiceType -> attributeKey.equals(choiceType.getAttribute().getName())
+                        || attributeKey.equals(choiceType.getAttribute().getFriendlyName()))
+                        // Several statements with same name are treated like one with several values
+                        .flatMap(choiceType -> choiceType.getAttribute().getAttributeValue().stream())
+                        .collect(Collectors.toList());
 
-            boolean attributeValueMatch = areAttributeValuesRegexes ? valueMatchesRegex(attribute.getValue(), attributeValues) : attributeValues.contains(attribute.getValue());
-            if (!attributeValueMatch) {
-                return false;
+                boolean attributeValueMatch = areAttributeValuesRegexes ? valueMatchesRegex(value, attributeValues) : attributeValues.contains(value);
+                if (!attributeValueMatch) {
+                    return false;
+                }
             }
         }
 

--- a/services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToRoleMapper.java
+++ b/services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToRoleMapper.java
@@ -17,7 +17,6 @@
 
 package org.keycloak.broker.saml.mappers;
 
-
 import org.keycloak.broker.provider.BrokeredIdentityContext;
 import org.keycloak.broker.provider.ConfigConstants;
 import org.keycloak.broker.saml.SAMLEndpoint;
@@ -122,8 +121,9 @@ public class AdvancedAttributeToRoleMapper extends AbstractAttributeToRoleMapper
         return "If the set of attributes exists and can be matched, grant the user the specified realm or client role.";
     }
 
+    @Override
     protected boolean applies(final IdentityProviderMapperModel mapperModel, final BrokeredIdentityContext context) {
-        Map<String, String> attributes = mapperModel.getConfigMap(ATTRIBUTE_PROPERTY_NAME);
+        Map<String, List<String>> attributes = mapperModel.getConfigMap(ATTRIBUTE_PROPERTY_NAME);
         boolean areAttributeValuesRegexes = Boolean.parseBoolean(mapperModel.getConfig().get(ARE_ATTRIBUTE_VALUES_REGEX_PROPERTY_NAME));
 
         AssertionType assertion = (AssertionType) context.getContextData().get(SAMLEndpoint.SAML_ASSERTION);
@@ -132,19 +132,21 @@ public class AdvancedAttributeToRoleMapper extends AbstractAttributeToRoleMapper
             return false;
         }
 
-        for (Map.Entry<String, String> attribute : attributes.entrySet()) {
-            String attributeKey = attribute.getKey();
-            List<Object> attributeValues = attributeAssertions.stream()
-                    .flatMap(statements -> statements.getAttributes().stream())
-                    .filter(choiceType -> attributeKey.equals(choiceType.getAttribute().getName())
-                            || attributeKey.equals(choiceType.getAttribute().getFriendlyName()))
-                    // Several statements with same name are treated like one with several values
-                    .flatMap(choiceType -> choiceType.getAttribute().getAttributeValue().stream())
-                    .collect(Collectors.toList());
+        for (Map.Entry<String, List<String>> entry : attributes.entrySet()) {
+            String attributeKey = entry.getKey();
+            for (String value : entry.getValue()) {
+                List<Object> attributeValues = attributeAssertions.stream()
+                        .flatMap(statements -> statements.getAttributes().stream())
+                        .filter(choiceType -> attributeKey.equals(choiceType.getAttribute().getName())
+                        || attributeKey.equals(choiceType.getAttribute().getFriendlyName()))
+                        // Several statements with same name are treated like one with several values
+                        .flatMap(choiceType -> choiceType.getAttribute().getAttributeValue().stream())
+                        .collect(Collectors.toList());
 
-            boolean attributeValueMatch = areAttributeValuesRegexes ? valueMatchesRegex(attribute.getValue(), attributeValues) : attributeValues.contains(attribute.getValue());
-            if (!attributeValueMatch) {
-                return false;
+                boolean attributeValueMatch = areAttributeValuesRegexes ? valueMatchesRegex(value, attributeValues) : attributeValues.contains(value);
+                if (!attributeValueMatch) {
+                    return false;
+                }
             }
         }
 

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedRoleMapperTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractAdvancedRoleMapperTest.java
@@ -27,6 +27,10 @@ public abstract class AbstractAdvancedRoleMapperTest extends AbstractRoleMapperT
             "    \"value\": \"value 1\"\n" +
             "  },\n" +
             "  {\n" +
+            "    \"key\": \"" + KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME + "\",\n" +
+            "    \"value\": \"value 2\"\n" +
+            "  },\n" +
+            "  {\n" +
             "    \"key\": \"" + KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2 + "\",\n" +
             "    \"value\": \"value 2\"\n" +
             "  }\n" +
@@ -152,7 +156,7 @@ public abstract class AbstractAdvancedRoleMapperTest extends AbstractRoleMapperT
         UserRepresentation user = findUser(bc.providerRealmName(), bc.getUserLogin(), bc.getUserEmail());
         ImmutableMap<String, List<String>> matchingAttributes = ImmutableMap.<String, List<String>> builder()
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME,
-                        ImmutableList.<String> builder().add("value 1").build())
+                        ImmutableList.<String> builder().add("value 1").add("value 2").build())
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2,
                         ImmutableList.<String> builder().add(newValueForAttribute2).build())
                 .put("some.other.attribute", ImmutableList.<String> builder().add("some value").build())
@@ -174,7 +178,7 @@ public abstract class AbstractAdvancedRoleMapperTest extends AbstractRoleMapperT
     private static Map<String, List<String>> createMatchingUserConfig() {
         return ImmutableMap.<String, List<String>> builder()
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME,
-                        ImmutableList.<String> builder().add("value 1").build())
+                        ImmutableList.<String> builder().add("value 1").add("value 2").build())
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2,
                         ImmutableList.<String> builder().add("value 2").build())
                 .build();

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractGroupBrokerMapperTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractGroupBrokerMapperTest.java
@@ -36,6 +36,10 @@ public abstract class AbstractGroupBrokerMapperTest extends AbstractGroupMapperT
             "    \"value\": \"value 1\"\n" +
             "  },\n" +
             "  {\n" +
+            "    \"key\": \"" + KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME + "\",\n" +
+            "    \"value\": \"value 2\"\n" +
+            "  },\n" +
+            "  {\n" +
             "    \"key\": \"" + KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2 + "\",\n" +
             "    \"value\": \"value 2\"\n" +
             "  }\n" +
@@ -69,7 +73,7 @@ public abstract class AbstractGroupBrokerMapperTest extends AbstractGroupMapperT
     protected void updateUser() {
         UserRepresentation user = findUser(bc.providerRealmName(), bc.getUserLogin(), bc.getUserEmail());
         ImmutableMap<String, List<String>> matchingAttributes = ImmutableMap.<String, List<String>>builder()
-                .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME, ImmutableList.<String>builder().add("value 1").build())
+                .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME, ImmutableList.<String>builder().add("value 1").add("value 2").build())
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2, ImmutableList.<String>builder().add(newValueForAttribute2).build())
                 .put("some.other.attribute", ImmutableList.<String>builder().add("some value").build())
                 .build();
@@ -112,7 +116,7 @@ public abstract class AbstractGroupBrokerMapperTest extends AbstractGroupMapperT
     protected static Map<String, List<String>> createMatchingAttributes() {
         return ImmutableMap.<String, List<String>> builder()
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME,
-                        ImmutableList.<String> builder().add("value 1").build())
+                        ImmutableList.<String> builder().add("value 1").add("value 2").build())
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2,
                         ImmutableList.<String> builder().add("value 2").build())
                 .build();

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToGroupMapperTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToGroupMapperTest.java
@@ -28,6 +28,9 @@ public class KcSamlAdvancedAttributeToGroupMapperTest extends AbstractGroupBroke
             "  {\n" +
             "    \"key\": \"" + ATTRIBUTE_TO_MAP_FRIENDLY_NAME + "\",\n" +
             "    \"value\": \"value 1\"\n" +
+            "  },\n" +"  {\n" +
+            "    \"key\": \"" + ATTRIBUTE_TO_MAP_FRIENDLY_NAME + "\",\n" +
+            "    \"value\": \"value 2\"\n" +
             "  },\n" +
             "  {\n" +
             "    \"key\": \"" + KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2 + "\",\n" +
@@ -65,7 +68,7 @@ public class KcSamlAdvancedAttributeToGroupMapperTest extends AbstractGroupBroke
     public void attributeFriendlyNameGetsConsideredAndMatchedToGroup() {
         createAdvancedGroupMapper(ATTRIBUTES, false,KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2);
         createUserInProviderRealm(ImmutableMap.<String, List<String>> builder()
-                .put(ATTRIBUTE_TO_MAP_FRIENDLY_NAME, ImmutableList.<String> builder().add("value 1").build())
+                .put(ATTRIBUTE_TO_MAP_FRIENDLY_NAME, ImmutableList.<String> builder().add("value 1").add("value 2").build())
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2,
                         ImmutableList.<String> builder().add("value 2").build())
                 .build());

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToRoleMapperTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToRoleMapperTest.java
@@ -26,6 +26,10 @@ public class KcSamlAdvancedAttributeToRoleMapperTest extends AbstractAdvancedRol
             "    \"value\": \"value 1\"\n" +
             "  },\n" +
             "  {\n" +
+            "    \"key\": \"" + ATTRIBUTE_TO_MAP_FRIENDLY_NAME + "\",\n" +
+            "    \"value\": \"value 2\"\n" +
+            "  },\n" +
+            "  {\n" +
             "    \"key\": \"" + KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2 + "\",\n" +
             "    \"value\": \"value 2\"\n" +
             "  }\n" +
@@ -58,7 +62,7 @@ public class KcSamlAdvancedAttributeToRoleMapperTest extends AbstractAdvancedRol
     public void attributeFriendlyNameGetsConsideredAndMatchedToRole() {
         createAdvancedRoleMapper(ATTRIBUTES, false);
         createUserInProviderRealm(ImmutableMap.<String, List<String>> builder()
-                .put(ATTRIBUTE_TO_MAP_FRIENDLY_NAME, ImmutableList.<String> builder().add("value 1").build())
+                .put(ATTRIBUTE_TO_MAP_FRIENDLY_NAME, ImmutableList.<String> builder().add("value 1").add("value 2").build())
                 .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2,
                         ImmutableList.<String> builder().add("value 2").build())
                 .build());

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/OidcClaimToUserSessionNoteMapperTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/OidcClaimToUserSessionNoteMapperTest.java
@@ -123,7 +123,7 @@ public class OidcClaimToUserSessionNoteMapperTest extends AbstractIdentityProvid
 
     @Test
     public void claimIsNotPropagatedWhenNameDoesNotMatch() {
-        createUserSessionNoteIdpMapper(IdentityProviderMapperSyncMode.IMPORT, "something-unexpected");
+        createUserSessionNoteIdpMapper(IdentityProviderMapperSyncMode.IMPORT, "something-unexpected-1", "something-unexpected-2");
 
         AccessToken accessToken = login();
 
@@ -157,7 +157,7 @@ public class OidcClaimToUserSessionNoteMapperTest extends AbstractIdentityProvid
     }
 
     private IdentityProviderMapperRepresentation createUserSessionNoteIdpMapper(IdentityProviderMapperSyncMode syncMode,
-            String matchingValue) {
+            String... matchingValue) {
         IdentityProviderMapperRepresentation mapper = new IdentityProviderMapperRepresentation();
         mapper.setName("User Session Note Idp Mapper");
         mapper.setIdentityProviderMapper(ClaimToUserSessionNoteMapper.PROVIDER_ID);
@@ -170,8 +170,17 @@ public class OidcClaimToUserSessionNoteMapperTest extends AbstractIdentityProvid
         return persistMapper(mapper);
     }
 
-    private String createClaimsConfig(String matchingValue) {
-        return "[{\"key\":\"" + CLAIM_NAME + "\",\"value\":\"" + matchingValue + "\"}]";
+    private String createClaimsConfig(String... matchingValue) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[");
+        if (matchingValue != null) {
+            for (String value : matchingValue) {
+                sb.append("{\"key\":\"").append(CLAIM_NAME).append("\",\"value\":\"").append(value).append("\"},");
+            }
+            sb.setLength(sb.length() - 1);
+        }
+        sb.append("]");
+        return sb.toString();
     }
 
     private void updateProviderHardcodedClaimMapper(String value) {

